src/shared/bpf-object.c

```diff
@@ -0,0 +1,139 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include "bpf-object.h"
+#include "bpf-program-v2.h"
+#include "fd-util.h"
+#include "set.h"
+
+int bpf_object_new(const unsigned char *mem_buf, size_t size, struct bpf_object **object) {
+        _cleanup_(bpf_object_freep) struct bpf_object *p = NULL;
+        /* Standard error code. */
+        long err;
+
+        assert(mem_buf);
+        assert(size > 0);
+        assert(object);
+
+        DECLARE_LIBBPF_OPTS(
+                        bpf_object_open_opts,
+                        opts);
+
+        p = bpf_object__open_mem(mem_buf, size, &opts);
+        /* Should either point to a valid BPF object or store standard error code. */
+        if (!p)
+                return -EINVAL;
+
+        err = libbpf_get_error(p);
+        if (err)
+                return -err;
+
+        *object = TAKE_PTR(p);
+        return 0;
+}
+
+struct bpf_object *bpf_object_free(struct bpf_object *object) {
+        if (object)
+                bpf_object__close(object);
+
+        return NULL;
+}
```

Comment:
I'm not familiar with the systemd code base, but why to return a pointer on a free operation?

Reply:
Common pattern we use to free and null in a single line: p = free_(p)

Reply:
@mauriciovasquezbernal it originates in the usual xyz_ref()/xyz_unref() functions used with ref-counting code, where most codebases make the xyz_ref() function return the same object that was passed in. And if you think about it a bit, then the xyz_unref() function should do the same, but return NULL (cast to the right type), since it invalidates the object (at least for the local caller). And then when you think about it some more you come to the conclusion that xyz_free() should actuallly do the same (i.e. for non-refcounted object allocation management).

Reply:
That makes perfect sense to me. Thank you both for the detailed explanation.